keps/sig-scheduling/3022-min-domains-in-pod-topology-spread/README.md

```diff
@@ -0,0 +1,506 @@
+In some cases, users want to limit the number of domains for the cluster autoscaler and force spreading Pods over a minimum number of domains.
+
+### Goals
+
+- Users can specify `minDomains` to limit the number of domains.
+- Users can use it as a mandatory requirement with `WhenUnsatisfiable=DoNotSchedule`.
+- Users can use it as a best-efforts manner with `WhenUnsatisfiable=ScheduleAnyway`.
+
+### Non-Goals
+
+- Add new field to limit the maximum number of topology domains.
+
+## Proposal
+
+### User Story
+
+I am using cluster autoscaler and I want to force spreading a deployment over at least 5 Nodes.
+
+## Design Details
+
+Users can define a minimum number of domains with `minDomains` parameter.
+
+Pod Topology Spread has the semantics of "global minimum", which means the minimum number of pods that match the label selector in a topology domain.
+
+When the number of domains that have matching Pods is less than `minDomains`,
+Pod Topology Spread treats "global minimum" as 0.
+
+As the result,
+
+- when `whenUnsatisfiable` equals `DoNotSchedule`, scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+- when `whenUnsatisfiable` equals `ScheduleAnyway`, scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+
+### API
+
+New optional parameter called `MinDomains` is introduced to `PodSpec.TopologySpreadConstraint[*]`.
+
+```go
+type TopologySpreadConstraint struct {
+......
+  // When the number of domains that have matching Pods is less than `minDomains`,
+  // Pod Topology Spread treats "global minimum" as 0.
+  // As the result,
+  // - when `whenUnsatisfiable` equals `DoNotSchedule`,
+  //   scheduler doesn't schedule a matching Pod to Nodes on the domains that have the same number of matching Pods as `maxSkew`.
+  // - when `whenUnsatisfiable` equals `ScheduleAnyway`,
+  //   scheduler prefers Nodes on the domains that don't have the same number of matching Pods as `maxSkew`.
+  // +optional
+  MinDomains *int32
+}
+```
+
+### How user stories are addressed
+
+Users can set `MinDomains` and `whenUnsatisfiable: DoNotSchedule` to achieve it.
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nginx-deployment
+spec:
+  selector:
+    matchLabels:
+      app: nginx
+  replicas: 15
+  template:
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+        - name: nginx
+          image: nginx:1.14.2
+          ports:
+            - containerPort: 80
+          topologySpreadConstraints:
+            - maxSkew: 2
+              minDomains: 5
+              topologyKey: kubernetes.io/hostname
+              whenUnsatisfiable: DoNotSchedule
+              labelSelector:
+                matchLabels:
+                  foo: bar
+```
+
+Until 10 Pods have been created, this flow is expected to repeats itself.
```

Comment:
This is assuming that the cluster initially only has 0 or 1 node.

Reply:
True. I changed the explanation a bit. How about this?